File: apps/slf-esport/frontend/src/types/session.ts
/**
 * Session types
 */

export enum SessionType {
  SOLO = 'solo',
  DUO = 'duo',
  TRIO = 'trio',
  TEAM = 'team',
  GROUP = 'group',
}

export enum SessionStatus {
  PENDING = 'pending',
  CONFIRMED = 'confirmed',
  CANCELLED = 'cancelled',
  COMPLETED = 'completed',
}

export interface SessionParticipant {
  id: number
  session_id: number
  user_id: number
  attendance_status: string
  response_status: string  // pending, confirmed, maybe, declined
  invitation_sent_at?: string
  response_at?: string
  decline_reason?: string
  notes?: string
  created_at: string
  updated_at: string
  // User info
  username?: string
  full_name?: string
  game_username?: string
}

export interface Session {
  id: number
  title: string
  description?: string
  session_type: SessionType
  status: SessionStatus
  start_time: string
  end_time: string
  duration_minutes?: number
  coach_id?: number
  created_by_id: number
  meeting_url?: string
  notes?: string
  created_at: string
  updated_at: string
  participants: SessionParticipant[]
}

export interface SessionCreate {
  title: string
  description?: string
  session_type: SessionType
  start_time: string
  end_time: string
  coach_id?: number
  meeting_url?: string
  notes?: string
  participant_ids?: number[]
}

export interface SessionUpdate {
  title?: string
  description?: string
  session_type?: SessionType
  start_time?: string
  end_time?: string
  status?: SessionStatus
  coach_id?: number
  meeting_url?: string
  notes?: string
}

export interface SessionStats {
  total_sessions: number
  completed_sessions: number
  upcoming_sessions: number
  cancelled_sessions: number
  attendance_rate: number
  total_hours: number
}

export interface SessionListResponse {
  total: number
  sessions: Session[]
  page: number
  page_size: number
}

// UI constants
export const SESSION_TYPE_LABELS: Record<SessionType, string> = {
  [SessionType.SOLO]: 'Solo',
  [SessionType.DUO]: 'Duo',
  [SessionType.TRIO]: 'Trio',
  [SessionType.TEAM]: 'Team',
  [SessionType.GROUP]: 'Groupe',
}

export const SESSION_TYPE_EMOJIS: Record<SessionType, string> = {
  [SessionType.SOLO]: '🎯',
  [SessionType.DUO]: '👥',
  [SessionType.TRIO]: '🔺',
  [SessionType.TEAM]: '⚔️',
  [SessionType.GROUP]: '👨‍👩‍👧‍👦',
}

export const SESSION_TYPE_COLORS: Record<SessionType, string> = {
  [SessionType.SOLO]: 'bg-blue-100 text-blue-800 dark:bg-blue-900/20 dark:text-blue-300',
  [SessionType.DUO]: 'bg-green-100 text-green-800 dark:bg-green-900/20 dark:text-green-300',
  [SessionType.TRIO]: 'bg-purple-100 text-purple-800 dark:bg-purple-900/20 dark:text-purple-300',
  [SessionType.TEAM]: 'bg-orange-100 text-orange-800 dark:bg-orange-900/20 dark:text-orange-300',
  [SessionType.GROUP]: 'bg-yellow-100 text-yellow-800 dark:bg-yellow-900/20 dark:text-yellow-300',
}

export const SESSION_STATUS_LABELS: Record<SessionStatus, string> = {
  [SessionStatus.PENDING]: 'En attente',
  [SessionStatus.CONFIRMED]: 'Confirmé',
  [SessionStatus.CANCELLED]: 'Annulé',
  [SessionStatus.COMPLETED]: 'Complété',
}

export const SESSION_STATUS_COLORS: Record<SessionStatus, string> = {
  [SessionStatus.PENDING]: 'bg-gray-100 text-gray-800 dark:bg-gray-800 dark:text-gray-300',
  [SessionStatus.CONFIRMED]: 'bg-green-100 text-green-800 dark:bg-green-900/20 dark:text-green-300',
  [SessionStatus.CANCELLED]: 'bg-red-100 text-red-800 dark:bg-red-900/20 dark:text-red-300',
  [SessionStatus.COMPLETED]: 'bg-blue-100 text-blue-800 dark:bg-blue-900/20 dark:text-blue-300',
}

// Response status labels (participant invitation response)
export const RESPONSE_STATUS_LABELS: Record<string, string> = {
  pending: 'En attente',
  confirmed: 'Confirmé',
  maybe: 'Peut-être',
  declined: 'Décliné',
}

export const RESPONSE_STATUS_COLORS: Record<string, string> = {
  pending: 'bg-yellow-100 text-yellow-800 dark:bg-yellow-900/20 dark:text-yellow-300',
  confirmed: 'bg-green-100 text-green-800 dark:bg-green-900/20 dark:text-green-300',
  maybe: 'bg-blue-100 text-blue-800 dark:bg-blue-900/20 dark:text-blue-300',
  declined: 'bg-red-100 text-red-800 dark:bg-red-900/20 dark:text-red-300',
}

File: apps/slf-esport/backend/app/schemas/session.py
"""
Session schemas - Pydantic validation for sessions
"""

from datetime import datetime
from typing import Optional, List
from pydantic import BaseModel, Field, validator, model_validator
from app.models.session import SessionType, SessionStatus


# ========== Session Participant Schemas ==========

class SessionParticipantBase(BaseModel):
    """Base session participant schema"""
    user_id: int
    attendance_status: str = "pending"
    notes: Optional[str] = None


class SessionParticipantCreate(SessionParticipantBase):
    """Schema for creating a session participant"""
    pass


class SessionParticipantUpdate(BaseModel):
    """Schema for updating a session participant"""
    attendance_status: Optional[str] = None
    notes: Optional[str] = None


class SessionParticipantResponse(SessionParticipantBase):
    """Schema for session participant response"""
    id: int
    session_id: int
    response_status: str = "pending"  # pending, confirmed, maybe, declined
    invitation_sent_at: Optional[datetime] = None
    response_at: Optional[datetime] = None
    decline_reason: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    # User info (populated from relationship)
    username: Optional[str] = None
    full_name: Optional[str] = None
    game_username: Optional[str] = None

    @model_validator(mode='before')
    @classmethod
    def populate_user_info(cls, data):
        """Populate user info from the relationship if available"""
        if hasattr(data, 'user') and data.user:
            # Convert SQLAlchemy model to dict and enrich with user data
            if isinstance(data, dict):
                return data
            else:
                # data is a SQLAlchemy model
                participant_dict = {
                    'id': data.id,
                    'session_id': data.session_id,
                    'user_id': data.user_id,
                    'attendance_status': data.attendance_status,
                    'notes': data.notes,
                    'response_status': data.response_status,
                    'invitation_sent_at': data.invitation_sent_at,
                    'response_at': data.response_at,
                    'decline_reason': data.decline_reason,
                    'created_at': data.created_at,
                    'updated_at': data.updated_at,
                    'username': data.user.username,
                    'full_name': data.user.full_name,
                    'game_username': data.user.game_username,
                }
                return participant_dict
        return data

    class Config:
        from_attributes = True


# ========== Session Schemas ==========

class SessionBase(BaseModel):
    """Base session schema"""
    title: str = Field(..., min_length=1, max_length=200)
    description: Optional[str] = None
    session_type: SessionType
    start_time: datetime
    end_time: datetime
    coach_id: Optional[int] = None
    meeting_url: Optional[str] = None
    notes: Optional[str] = None

    @validator('end_time')
    def validate_end_time(cls, v, values):
        """Ensure end_time is after start_time"""
        if 'start_time' in values and v <= values['start_time']:
            raise ValueError('end_time must be after start_time')
        return v


class SessionCreate(SessionBase):
    """Schema for creating a session"""
    participant_ids: Optional[List[int]] = Field(default_factory=list)


class SessionUpdate(BaseModel):
    """Schema for updating a session"""
    title: Optional[str] = Field(None, min_length=1, max_length=200)
    description: Optional[str] = None
    session_type: Optional[SessionType] = None
    start_time: Optional[datetime] = None
    end_time: Optional[datetime] = None
    status: Optional[SessionStatus] = None
    coach_id: Optional[int] = None
    meeting_url: Optional[str] = None
    notes: Optional[str] = None


class SessionResponse(SessionBase):
    """Schema for session response"""
    id: int
    status: SessionStatus
    duration_minutes: Optional[int]
    created_by_id: int
    created_at: datetime
    updated_at: datetime
    participants: List[SessionParticipantResponse] = []

    class Config:
        from_attributes = True


class SessionListResponse(BaseModel):
    """Schema for session list with pagination"""
    total: int
    sessions: List[SessionResponse]
    page: int
    page_size: int


# ========== Session Stats ==========

class SessionStats(BaseModel):
    """Session statistics for a user"""
    total_sessions: int
    completed_sessions: int
    upcoming_sessions: int
    cancelled_sessions: int
    attendance_rate: float  # % of sessions attended
    total_hours: float  # Total training hours
